File: src/utils/dailyMetrics.js
import {
  calcBMR,
  calcTrainingKcal,
  getTodayKey,
  getTodayStr,
  roundTo,
  toNullableNumber,
} from './calcBase.js'

function getRestPlan() {
  return { type: 'rest', exercises: [] }
}

function calcBMI(profileBasic = {}) {
  const weight = toNullableNumber(profileBasic.weight)
  const heightCm = toNullableNumber(profileBasic.height)

  if (weight === null || weight <= 0 || heightCm === null || heightCm <= 0) {
    return null
  }

  const heightM = heightCm / 100
  return roundTo(weight / (heightM * heightM), 1)
}

// 蛋白质按体重折算采用 g/kg，适合作为 prompt 与 UI 的统一口径。
function calcProteinPerKg(proteinIntake, weight) {
  if (!Number.isFinite(proteinIntake) || proteinIntake <= 0) {
    return null
  }

  if (!Number.isFinite(weight) || weight <= 0) {
    return null
  }

  return roundTo(proteinIntake / weight, 1)
}

function getCalorieStatus(delta) {
  if (!Number.isFinite(delta)) {
    return 'unknown'
  }

  if (delta > 100) {
    return 'surplus'
  }

  if (delta < -100) {
    return 'deficit'
  }

  return 'balanced'
}

function getProteinStatus(proteinPerKg) {
  if (!Number.isFinite(proteinPerKg)) {
    return 'unknown'
  }

  return proteinPerKg >= 1.6 ? 'met' : 'low'
}

// 汇总今日本地即可确定的训练、营养与恢复指标，供 prompt 注入和页面展示复用。
export function buildDailyMetricsSummary(
  profile = {},
  weeklyPlan = {},
  dailyLog = {},
  referenceDate,
) {
  const todayKey = getTodayKey(referenceDate)
  const todayStr = getTodayStr(referenceDate)
  const todayPlan = weeklyPlan?.[todayKey] ?? getRestPlan()
  const isTrainingDay = todayPlan.type !== 'rest'
  const bmr = calcBMR(profile.basic)
  const trainingKcal = isTrainingDay
    ? calcTrainingKcal(todayPlan.exercises, profile.oneRM)
    : 0
  const tdee = Math.round(bmr * 1.2 + trainingKcal)
  const todayLog = dailyLog?.[todayStr] ?? {}
  const calorieIntake = toNullableNumber(todayLog.kcal)
  const proteinIntake = toNullableNumber(todayLog.protein)
  const sleepHours = toNullableNumber(todayLog.sleep)
  const fatigueLevel = toNullableNumber(todayLog.fatigue)
  const calorieDelta = calorieIntake === null ? null : calorieIntake - tdee
  const bodyWeight = toNullableNumber(profile?.basic?.weight)
  const proteinPerKg = calcProteinPerKg(proteinIntake, bodyWeight)

  return {
    todayKey,
    todayStr,
    todayPlan,
    isTrainingDay,
    bmr,
    bmi: calcBMI(profile.basic),
    trainingKcal,
    tdee,
    calorie: {
      intake: calorieIntake,
      delta: calorieDelta,
      status: getCalorieStatus(calorieDelta),
    },
    protein: {
      intake: proteinIntake,
      gramsPerKg: proteinPerKg,
      status: getProteinStatus(proteinPerKg),
    },
    recovery: {
      sleepHours,
      fatigueLevel,
    },
  }
}

// 保持现有 prompt 等调用方的返回结构，避免收口阶段扩大改动面。
export function calcTDEE(profile = {}, weeklyPlan = {}, dailyLog = {}, referenceDate) {
  const summary = buildDailyMetricsSummary(profile, weeklyPlan, dailyLog, referenceDate)

  return {
    todayKey: summary.todayKey,
    todayStr: summary.todayStr,
    todayPlan: summary.todayPlan,
    isTrainingDay: summary.isTrainingDay,
    bmr: summary.bmr,
    trainingKcal: summary.trainingKcal,
    tdee: summary.tdee,
    todayKcal: summary.calorie.intake ?? 0,
    delta: summary.calorie.delta ?? 0,
  }
}

File: backend/metrics/daily_metrics.py
from __future__ import annotations

from datetime import date
from typing import Any

from backend.db.models import WEEKDAY_ORDER


def build_daily_metrics_summary(
    *,
    profile: dict[str, Any] | None,
    weekly_plan: dict[str, Any],
    daily_log: dict[str, Any] | None,
    target_date: str,
) -> dict[str, Any]:
    today_key = _weekday_key(target_date)
    today_plan = weekly_plan.get(today_key, {"type": "rest", "exercises": []})
    is_training_day = today_plan.get("type") != "rest"
    basic = (profile or {}).get("basic") or {}
    one_rm = (profile or {}).get("oneRm") or {}
    bmr = _calc_bmr(basic)
    training_kcal = _calc_training_kcal(today_plan.get("exercises") or [], one_rm) if is_training_day else 0
    estimated_tdee = round(bmr * 1.2 + training_kcal) if bmr is not None else None
    manual_tdee = _to_nullable_number((daily_log or {}).get("tdeeManual"))
    tdee = manual_tdee if manual_tdee is not None else estimated_tdee
    calorie_intake = _to_nullable_number((daily_log or {}).get("kcal"))
    calorie_delta = calorie_intake - tdee if calorie_intake is not None and tdee is not None else None
    weight = _to_nullable_number(basic.get("weight"))
    protein_intake = _to_nullable_number((daily_log or {}).get("protein"))
    protein_per_kg = _round_to(protein_intake / weight, 1) if protein_intake and weight else None

    # 后端指标服务是后续统一展示、prompt 注入和 Agent 工具判断口径的来源。
    return {
        "today_key": today_key,
        "today_str": target_date,
        "today_plan_type": today_plan.get("type") or "rest",
        "is_training_day": is_training_day,
        "bmr_kcal": bmr,
        "training_kcal": training_kcal,
        "estimated_tdee_kcal": estimated_tdee,
        "tdee_source": "manual" if manual_tdee is not None else "estimated",
        "tdee_kcal": tdee,
        "bmi": _calc_bmi(basic),
        "steps_count": _to_nullable_number((daily_log or {}).get("steps")),
        "calorie_intake_kcal": calorie_intake,
        "calorie_delta_kcal": calorie_delta,
        "calorie_status": _calorie_status(calorie_delta),
        "protein_intake_g": protein_intake,
        "protein_g_per_kg": protein_per_kg,
        "protein_status": _protein_status(protein_per_kg),
        "sleep_hours": _to_nullable_number((daily_log or {}).get("sleep")),
        "fatigue_level": _to_nullable_number((daily_log or {}).get("fatigue")),
    }


def _weekday_key(target_date: str) -> str:
    parsed = date.fromisoformat(target_date)
    return WEEKDAY_ORDER[parsed.weekday()]


def _to_number(value: Any) -> float:
    try:
      parsed = float(value)
    except (TypeError, ValueError):
      return 0.0
    return parsed


def _to_nullable_number(value: Any) -> float | None:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return None
    return parsed


def _round_to(value: float, digits: int = 1) -> float:
    return round(value, digits)


def _calc_bmr(basic: dict[str, Any]) -> int | None:
    weight = _to_nullable_number(basic.get("weight"))
    height = _to_nullable_number(basic.get("height"))
    age = _to_nullable_number(basic.get("age"))
    if weight is None or height is None or age is None:
        return None
    offset = -161 if basic.get("sex") == "female" else 5
    return round(10 * weight + 6.25 * height - 5 * age + offset)


def _calc_bmi(basic: dict[str, Any]) -> float | None:
    weight = _to_nullable_number(basic.get("weight"))
    height = _to_nullable_number(basic.get("height"))
    if weight is None or height is None or height <= 0:
        return None
    height_m = height / 100
    return _round_to(weight / (height_m * height_m), 1)


def _calc_training_kcal(exercises: list[dict[str, Any]], one_rm: dict[str, Any]) -> int:
    total_volume = 0.0
    for exercise in exercises:
        kg = _exercise_kg(exercise, one_rm)
        total_volume += kg * _to_number(exercise.get("sets")) * _to_number(exercise.get("reps"))
    return round(total_volume * 0.1)


def _exercise_kg(exercise: dict[str, Any], one_rm: dict[str, Any]) -> float:
    ref = exercise.get("ref1RM")
    if ref:
        return _to_number(one_rm.get(ref)) * _to_number(exercise.get("pct"))
    return _to_number(exercise.get("kg"))


def _calorie_status(delta: float | None) -> str:
    if delta is None:
        return "unknown"
    if delta > 100:
        return "surplus"
    if delta < -100:
        return "deficit"
    return "balanced"


def _protein_status(protein_per_kg: float | None) -> str:
    if protein_per_kg is None:
        return "unknown"
    return "met" if protein_per_kg >= 1.6 else "low"
